enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,554 @@
+
+## Implementation History
+
+* This has already been implemented in Cluster API using an annotation based approach.
+
+## Drawbacks
+
+* If misused, this could prevent Machines from being deleted and lead to an increase in support
+cases to determine if it is safe to override the feature.
+
+## Alternatives
+
+### Custom Machine Controller
+Require advanced users to fork and customize.
+This can already be done if someone chooses, so not much of a solution.
+
+### Finalizers
+We define additional finalizers, but this really only implies the deletion lifecycle point.
+
+A misbehaving controller that accidentally removes finalizers could have undesirable effects,
+such as the removal of the Machine before it has been terminated.
+A misbehaving controller may also remove too many deletion hooks, but in this case, this
+wouldn't result in the Machine resource being removed from etcd and therefore prevents costly
+errors such as leaked cloud instances.
+
+The proposal today effectively extends the finalizer concept to allow two stages of finalization
+(pre and post drain) in conjunction with the existing Machine finalizer.
+
+A limitation of finalizers is the 63 character limit. With this limit, it would be difficult to
+provide information such as the owner of the deletion hook. By leveraging annotations which have
+a larger length limit, we can store more information about the hook within the API.
+
+If we were to leverage finalizers for this purpose, there are a few different ways this could work,
+as set out below.
+
+#### Machine API waits on all Finalizers
+
+In this scenario, the Machine API would not start draining/removing a Machine until all finalizers
+other than its own one have been removed. This would allow future controllers to manage the removal
+of the Machine without having Machine API have to know anything about their specific implementation.
+However, there could be a scenario where another controller also needs to wait for other finalizers
+to be removed, in such a case, unless the Machine API or this new controller are aware of each other
+there would be a deadlock created.
+
+Relying on all finalizers being removed before draining/removing could become fragile in the future
+due to the above, and as such is not a preferred solution.
+
+#### Machine API waits on some Finalizers
+
+In this scenario, Machine API would have a list of known finalizers that would prevent it from taking
+certain actions. For example a known etcd-quorum finalizer could prevent the Machine controller from
+draining the Machine.
+
+This creates a tight coupling between the Machine controller and other components within OpenShift.
+The upstream solution today, matches the proposal set out in this document. As we are experimenting
+with Cluster API in HyperShift, Central Machine Management and for future IPI clusters, it is
+preferred to not create a tight coupling with these components and to copy the upstream solution to
```

Comment:
To clarify, none of these fields exists upstream today, though nothing preclude us from proposing this approach. A hook mechanism is to be discussed as part of https://github.com/kubernetes-sigs/cluster-api/discussions/5556#discussioncomment-1575309